### PR TITLE
feat(commands): add leaderboard command

### PR DESCRIPTION
**Pull request by MixailE**
add command for info about balance top 10 users

### CODE REVIEW

**Comment by MixailE** [comment]
merge feat(commands): add leaderboard command

File: src/commands/leaderboard.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
+// Dependencies
+import { Telegraf, ContextMessageUpdate } from 'telegraf'
+import { User } from '../models/user';
```
Comment (backmeupplz) [inline src/commands/leaderboard.ts]: Need to remove `;`

File: src/commands/leaderboard.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+import { getLeaderBalance } from '../models'
+import { getName } from '../helpers/name'
+
+interface UserForLeaderBoard {
```
Comment (backmeupplz) [inline src/commands/leaderboard.ts]: Better to rename to `LeaderBoardUser`

File: src/commands/leaderboard.ts [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    username: String,
+    balance: Number,
+}
+
+const getUser = async (ctx: any, user: User): Promise<UserForLeaderBoard> => {
+    const member = await ctx.telegram.getChatMember(user.chatId, user.chatId);
+    return {
+        username: getName(member),
```
Comment (backmeupplz) [inline src/commands/leaderboard.ts]: Please, keep indentation to 2 spaces

File: src/commands/leaderboard.ts [hunk 4]
```diff
@@ -0,0 +1,36 @@
+// Dependencies
+import { Telegraf, ContextMessageUpdate } from 'telegraf'
+import { User } from '../models/user';
+import { getLeaderBalance } from '../models'
+import { getName } from '../helpers/name'
+
+interface UserForLeaderBoard {
+    username: String,
+    balance: Number,
```
Comment (backmeupplz) [inline src/commands/leaderboard.ts]: Better to use `number`

File: src/commands/leaderboard.ts [hunk 5]
```diff
@@ -0,0 +1,36 @@
+// Dependencies
+import { Telegraf, ContextMessageUpdate } from 'telegraf'
+import { User } from '../models/user';
+import { getLeaderBalance } from '../models'
+import { getName } from '../helpers/name'
+
+interface UserForLeaderBoard {
+    username: String,
```
Comment (backmeupplz) [inline src/commands/leaderboard.ts]: Better to use `string`

File: src/commands/leaderboard.ts [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    balance: Number,
+}
+
+const getUser = async (ctx: any, user: User): Promise<UserForLeaderBoard> => {
```
Comment (backmeupplz) [inline src/commands/leaderboard.ts]: 1. Remove return time (it will be inferred)
2. `ctx` should be `ContextMessageUpdate`

Comment (backmeupplz) [inline src/commands/leaderboard.ts]: Functions should be declared as `function`, not as `const`

File: src/commands/leaderboard.ts [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+    return {
+        username: getName(member),
+        balance: user.balance,
+    };
+}
+
+// Help commands
+export function setupLeaderBoard(bot: Telegraf<ContextMessageUpdate>) {
+  bot.command('leaderboard', async (ctx) => {
+    const countUserInLeader = 10;
```
Comment (backmeupplz) [inline src/commands/leaderboard.ts]: No need for extra variable, please just hardcode it to `countUserInLeader`

File: src/commands/leaderboard.ts [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+    // Get leaderboard users
```
Comment (backmeupplz) [inline src/commands/leaderboard.ts]: Please, change to `// Get users leaderboard`

File: src/models/user.ts [hunk 9]
```diff
@@ -19,4 +19,8 @@ export async function getUser(chatId: number) {
     user = await user.save()
   }
   return user
+}
+
+export function getLeaderBalance(chatId: number, limit: number) {
```
Comment (backmeupplz) [inline src/models/user.ts]: Please, rename to `getLeaderboard`

File: src/models/user.ts [hunk 10]
```diff
@@ -19,4 +19,8 @@ export async function getUser(chatId: number) {
...
+  return UserModel.find().sort({ balance: 'desc' }).limit(limit).exec();
```
Comment (backmeupplz) [inline src/models/user.ts]: Should be no need to use `exec` in the end

**Review by backmeupplz — CHANGES_REQUESTED** [review]
Please, make sure you use `Leaderboard` instead of `LeaderBoard`

File: src/commands/leaderboard.ts [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+import { User } from '../models/user'
+import { getLeaderboard } from '../models'
+import { getName } from '../helpers/name'
+
+interface LeaderBoardUser {
+  string: String,
+  number: Number,
+}
+
+async function getUser(ctx: any, user: User): Promise<LeaderBoardUser> {
```
Comment (backmeupplz) [inline src/commands/leaderboard.ts]: Please, remove return type

File: src/commands/leaderboard.ts [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+  const member = await ctx.telegram.getChatMember(user.chatId, user.chatId)
+  return {
+    string: getName(member),
+    number: user.balance,
+  }
```
Comment (backmeupplz) [inline src/commands/leaderboard.ts]: Add `as LeaderBoardUser`

File: src/commands/leaderboard.ts [hunk 13]
```diff
@@ -0,0 +1,35 @@
+// Dependencies
+import { Telegraf, ContextMessageUpdate } from 'telegraf'
+import { User } from '../models/user'
+import { getLeaderboard } from '../models'
+import { getName } from '../helpers/name'
+
+interface LeaderBoardUser {
```
Comment (backmeupplz) [inline src/commands/leaderboard.ts]: Rename to `LeaderboardUser`

File: src/commands/leaderboard.ts (hunk 11 above)
Comment (backmeupplz) [inline src/commands/leaderboard.ts]: `ctx` argument should be of `ContextMessageUpdate` type

File: src/commands/leaderboard.ts [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+  bot.command('leaderboard', async (ctx) => {
+    // Get users leaderboard
+    const users = await getLeaderboard()
+    // Get chat users
+    const members = await Promise.all(users.map((user: User) => getUser(ctx, user)))
+
+    // Prepare leaderboard
+    const list = members.map((member: LeaderBoardUser) => `*${member.string}*: *${member.number}*`).join('\n')
+    // Prepare text
+    const text = `🏆  Топ Мемолиархов 🏆\n\n${list}`
```
Comment (backmeupplz) [inline src/commands/leaderboard.ts]: Remove extra space after first `🏆`

File: src/commands/leaderboard.ts [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+  }
+}
+
+// Help commands
+export function setupLeaderBoard(bot: Telegraf<ContextMessageUpdate>) {
+  bot.command('leaderboard', async (ctx) => {
+    // Get users leaderboard
+    const users = await getLeaderboard()
+    // Get chat users
+    const members = await Promise.all(users.map((user: User) => getUser(ctx, user)))
```
Comment (backmeupplz) [inline src/commands/leaderboard.ts]: `members` should have a distinct type here. Add `: LeaderboardUser[]` here.

File: src/commands/leaderboard.ts [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+
+    // Prepare leaderboard
+    const list = members.map((member: LeaderBoardUser) => `*${member.string}*: *${member.number}*`).join('\n')
```
Comment (backmeupplz) [inline src/commands/leaderboard.ts]: Remove casting a type (`: LeaderBoardUser`) from here — it should be inferred from the `members` type

File: src/commands/leaderboard.ts [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+
+interface LeaderBoardUser {
+  string: String,
+  number: Number,
+}
+
+async function getUser(ctx: any, user: User): Promise<LeaderBoardUser> {
+  const member = await ctx.telegram.getChatMember(user.chatId, user.chatId)
+  return {
+    string: getName(member),
```
Comment (backmeupplz) [inline src/commands/leaderboard.ts]: Rename to `name`

File: src/commands/leaderboard.ts [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+    number: user.balance,
```
Comment (backmeupplz) [inline src/commands/leaderboard.ts]: Rename to `balance`

**Review by backmeupplz — CHANGES_REQUESTED** [review]

File: src/commands/leaderboard.ts [hunk 19]
```diff
@@ -0,0 +1,34 @@
...
+async function getUser(ctx: ContextMessageUpdate, user: User) {
+  const member = await ctx.telegram.getChatMember(user.chatId, user.chatId)
+  return ({
+    name: getName(member),
+    balance: user.balance,
+  } as LeaderboardUser)
+}
+
+// Help commands
+export function setupLeaderBoard(bot: Telegraf<ContextMessageUpdate>) {
```
Comment (backmeupplz) [inline src/commands/leaderboard.ts]: Rename to `setupLeaderboard`

File: src/commands/leaderboard.ts [hunk 20]
```diff
@@ -0,0 +1,34 @@
...
+import { getName } from '../helpers/name'
+
+interface LeaderboardUser {
+  name: String,
+  balance: Number,
+}
+
+async function getUser(ctx: ContextMessageUpdate, user: User) {
+  const member = await ctx.telegram.getChatMember(user.chatId, user.chatId)
+  return ({
```
Comment (backmeupplz) [inline src/commands/leaderboard.ts]: Remove `()`

File: src/commands/leaderboard.ts [hunk 21]
```diff
@@ -0,0 +1,34 @@
...
+    name: getName(member),
+    balance: user.balance,
+  } as LeaderboardUser)
+}
+
+// Help commands
+export function setupLeaderBoard(bot: Telegraf<ContextMessageUpdate>) {
+  bot.command('leaderboard', async (ctx: ContextMessageUpdate) => {
```
Comment (backmeupplz) [inline src/commands/leaderboard.ts]: Remove `:ContextMessageUpdate`, it's inferred

File: src/commands/leaderboard.ts [hunk 22]
```diff
@@ -0,0 +1,34 @@
...
+    // Get users leaderboard
+    const users = await getLeaderboard()
+    // Get chat users
+    const members: LeaderboardUser[] = await Promise.all(users.map((user: User) => getUser(ctx, user)))
```
Comment (backmeupplz) [inline src/commands/leaderboard.ts]: Remove `: User`, it's inferred

File: src/models/user.ts [hunk 23]
```diff
@@ -19,4 +19,8 @@ export async function getUser(chatId: number) {
     user = await user.save()
   }
   return user
+}
+
+export function getLeaderboard() {
```
Comment (backmeupplz) [inline src/models/user.ts]: Add `async`

**Review by backmeupplz — CHANGES_REQUESTED** [review]

**Pull request merged**